modules/miner/blockmanager.go

```diff
@@ -69,21 +69,53 @@ func (m *Miner) HeaderForWork() (types.BlockHeader, types.Target) {
 	lockID := m.mu.Lock()
 	defer m.mu.Unlock(lockID)
 
-	// Grab a block for work.
-	block, target := m.blockForWork()
+	var header types.BlockHeader
+	var randTxn types.Transaction
+	var block *types.Block
 
-	// Save a mapping between the block and its header, replacing the block
-	// that was stored 'headerForWorkMemory' requests ago.
+	if m.memProgress%headersPerBlockMemory == 0 {
+		// Grab a new block
+		block = new(types.Block)
+		*block, _ = m.blockForWork()
```

Comment:
If I don't allocate new memory, block won't be pointing to any memory, so it will seg fualt when I try to copy the new block into *block. 

Another way to do it would be:
`tempBlock, _ := m.blockForWork()`
`block = &tempBlock`